### ISSUE TITLE
ServerSide Row Model with pagination does not work

### ISSUE THREAD

**Issue by zhenyab** [ag-grid/ag-grid#3382]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/ceolter/ag-grid-angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Error appears in console whenever I try to use pagination and it does not call getRows of data source object

```
ERROR TypeError: Cannot read property 'rowIndex' of undefined
    at ServerSideCache.push.../../node_modules/ag-grid-enterprise/dist/lib/rowModels/serverSide/serverSideCache.js.ServerSideCache.getTopLevelRowDisplayedIndex (serverSideCache.js:292)
    at ServerSideRowModel.push.../../node_modules/ag-grid-enterprise/dist/lib/rowModels/serverSide/serverSideRowModel.js.ServerSideRowModel.getTopLevelRowDisplayedIndex (serverSideRowModel.js:384)
    at PaginationProxy.push.../../node_modules/ag-grid-community/dist/lib/rowModels/paginationProxy.js.PaginationProxy.calculatePagesMasterRowsOnly (paginationProxy.js:249)
    at PaginationProxy.push.../../node_modules/ag-grid-community/dist/lib/rowModels/paginationProxy.js.PaginationProxy.calculatePages (paginationProxy.js:198)
    at PaginationProxy.push.../../node_modules/ag-grid-community/dist/lib/rowModels/paginationProxy.js.PaginationProxy.onModelUpdated (paginationProxy.js:66)
    at PaginationProxy.push.../../node_modules/ag-grid-community/dist/lib/rowModels/paginationProxy.js.PaginationProxy.goToPage (paginationProxy.js:95)
    at PaginationProxy.push.../../node_modules/ag-grid-community/dist/lib/rowModels/paginationProxy.js.PaginationProxy.goToNextPage (paginationProxy.js:165)
    at PaginationComp.push.../../node_modules/ag-grid-community/dist/lib/rowModels/pagination/paginationComp.js.PaginationComp.onBtNext (paginationComp.js:102)
    at ZoneDelegate.invokeTask (zone-evergreen.js:391)
    at Object.onInvokeTask (ng_zone.ts:262)
    at ZoneDelegate.invokeTask (zone-evergreen.js:390)
    at Zone.runTask (zone-evergreen.js:168)
    at ZoneTask.invokeTask [as invoke] (zone-evergreen.js:465)
    at invokeTask (zone-evergreen.js:1603)
    at HTMLDivElement.globalZoneAwareCallback (zone-evergreen.js:1629)
```

**Expected behavior**
Pagination should call getRows and as user I should see the new results of the query in the table

**Minimal reproduction of the problem with instructions**

Here I provide the code I currently use with ag-grid
```
@Component({
  selector: 'search-page-grid',
  templateUrl: './search-page-grid.component.html',
  styleUrls: ['./search-page-grid.component.scss'],
  encapsulation: ViewEncapsulation.None,
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class SearchPageGridComponent implements OnDestroy, AfterViewInit {
  private fieldType = {
    id: 'id',
    name: 'name',
    assignee: 'workflow.assignee.name',
    deadline: 'workflow.deadline',
    type: 'type.value',
    mime: 'previews.master.mimetype',
    modifiedAt: 'modifiedAt',
    workflowStep: 'workflow.workflowStep.value'
  };
  public gridOptions: GridOptions = {
    columnDefs: [
      { headerName: 'ID', field: this.fieldType.id },
      { headerName: 'Name', field: this.fieldType.name },
      { headerName: 'Mime Type', field: this.fieldType.mime, sortable: false },
      { headerName: 'Type', field: this.fieldType.type },
      { headerName: 'Workflow step', field: this.fieldType.workflowStep },
      { headerName: 'Assigned to', field: this.fieldType.assignee, sortable: false },
      {
        headerName: 'Deadline',
        field: this.fieldType.deadline,
        cellRenderer: (data: ICellRendererParams) => this._deadlineCellRenderer(data.data),
      },
      {
        headerName: 'Modification',
        field: this.fieldType.modifiedAt,
        cellRenderer: (data: ICellRendererParams) => this._modificationCellRenderer(data.data),
      },
    ],
    pagination: true,
    paginationPageSize: 5,
    cacheBlockSize: 5,
    infiniteInitialRowCount: 5,
    maxBlocksInCache: 1,
    rowModelType: 'serverSide',
    enableSorting: true,
    enableServerSideSorting: true,
    enableServerSideFilter: true,
    serverSideSortingAlwaysResets: true
  };

  public totalAssetsCount$: Subject<number | null> = new Subject();

  selectedAssets: SearchAsset[] = [];

  private _needle = '';
  private _searchSubscription!: Subscription;
  private _deadlineCellRenderer = (data: SearchAsset) =>
    data && data.workflow.deadline && moment(data.workflow.deadline).format('L');
  private _modificationCellRenderer = (data: SearchAsset) =>
    data && data.modifiedAt && moment(data.modifiedAt).format('L LT');
  private _dataSource: IServerSideDatasource;

  public onGridReady(agGridEvent: AgGridEvent): void {
    agGridEvent.api.setServerSideDatasource(this._dataSource);
  }
  constructor(
    private _searchPageService: SearchPageService,
    private _searchService: SearchService,
    private _router: Router,
  ) {
    this._searchService.changeTerm$.subscribe(
      (term: string) => {
        this._needle = term;
        if (this.gridOptions.api) {
          this.gridOptions.api.setServerSideDatasource(this._dataSource);
        }
      },
      (error: any) => {
        console.log(error);
      },
    );

    const sortModelDictionary: any = this._getSortModelDictionary();
    this._dataSource = {
      getRows: (params: IServerSideGetRowsParams) => {

        console.log('PARAMS:', params.request);

        let cursor: Cursor = {
          after: params.request.startRow + '',
          limit: this.gridOptions.paginationPageSize,
        };
        let sortInfo = null;
        if (params.request.sortModel && params.request.sortModel.length > 0) {
          let sortCol = params.request.sortModel[0].colId;
          let sortDir = params.request.sortModel[0].sort;
          sortInfo = sortModelDictionary[sortCol];
          sortInfo.sortDirection = sortDir;
        }
        this._searchPageService.doQuery(this._needle, cursor, sortInfo).subscribe((result: SearchResults | null) => {
          if (result) {
            const assets: any[] = [...result.assets];
            this.totalAssetsCount$.next(result.info.totalCount);
            params.successCallback(assets, result.info.totalCount);
          }
        });
      }
    };

    this.totalAssetsCount$.next(0);
  }

  public onSelectionChanged(): void {
    if (this.gridOptions.api) {
      const selectedRows = this.gridOptions.api.getSelectedRows();
      this.selectedAssets = selectedRows;
    }
  }

  public onRowDblClick(event: RowEvent): void {
    this._router.navigateByUrl(`/${event.data.type.value.toLowerCase()}/${event.data.id}`);
  }

  public ngAfterViewInit(): void {
    if (this.gridOptions.api) {
      this.gridOptions.api.sizeColumnsToFit();
      this.gridOptions.api.hideOverlay();
    }
  }

  public ngOnDestroy(): void {
    if (this._searchSubscription) {
      this._searchSubscription.unsubscribe();
    }
  }

  private _getSortModelDictionary(): any {
    var sortModelDictionary: any = {};
    sortModelDictionary[this.fieldType.name] = { dataType: 'string', trait: 'display', field: 'name' };
    sortModelDictionary[this.fieldType.type] = { dataType: 'string', trait: 'type', field: 'type' };
    sortModelDictionary[this.fieldType.workflowStep] = { dataType: 'string', trait: 'workflow', field: 'workflowStep' };
    sortModelDictionary[this.fieldType.deadline] = { dataType: 'date', trait: 'workflow', field: 'deadline' };
    sortModelDictionary[this.fieldType.id] = { dataType: 'integer', trait: 'ids', field: 'id' };
    sortModelDictionary[this.fieldType.modifiedAt] = { dataType: 'dateTime', trait: 'modified', field: 'date' };
    return sortModelDictionary;
  }
}
```

Template HTML
```
<div class="grid search-page-grid-height">
  <div class="grid__col-8 search-details-grid">
    <ag-grid-angular class="ag-theme-material search-details-grid" (selectionChanged)="onSelectionChanged()"
      rowSelection="multiple" (rowDoubleClicked)="onRowDblClick($event)" (gridReady)="onGridReady($event)"
      [gridOptions]="gridOptions">
    </ag-grid-angular>
  </div>
  <div class="grid__col details-col">
    <search-details-panel class="search-details-panel" [selectedAssets]="selectedAssets"></search-details-panel>
  </div>
</div>
```

**Please tell us about your environment:**

MacOS X Mojave 10.4.6 
NodeJS 10.16.0 
Yarn 1.17.3 

* **ag-Grid/ag-Grid-Enterprise version:** 2.0.X

ag-grid-enterprise 21.2.0 

* **Angular version:** 2.0.X

angular 8.2.4

**Comment by gps-dev** [comment]
same problem here

**Comment by mkljakubowski** [comment]
Which is the latest version that works?

**Comment by lgorbliukxappia** [comment]
Same problem here.
This is even reproducible on the official test site under certain conditions:
https://www.ag-grid.com/example-runner/react.php?section=javascript-grid-server-side-model-pagination&example=pagination&lodash=1&enterprise=1&generated=1&grid=%7B%22theme%22%3A%22ag-theme-balham%22%2C%22noStyle%22%3A0%2C%22height%22%3A%22100%25%22%2C%22width%22%3A%22100%25%22%2C%22enterprise%22%3A1%7D
And it only happends depending on the amount of results. If you change the size of the page so that only 4, or 5, or 10 results are displayed, and you start changing the page number, eventually when you want to see past 100 results the issue starts happening.
On Firefox the error is:
### TypeError: lastRowNode is undefined
On Chrome:
### serverSideCache.js:292 Uncaught TypeError: Cannot read property 'rowIndex' of undefined
    at ServerSideCache.getTopLevelRowDisplayedIndex (serverSideCache.js:292)
    at ServerSideRowModel.getTopLevelRowDisplayedIndex (serverSideRowModel.js:384)
    at PaginationProxy.calculatePagesMasterRowsOnly (ag-grid-community.js:21137)
    at PaginationProxy.calculatePages (ag-grid-community.js:21093)
    at PaginationProxy.onModelUpdated (ag-grid-community.js:20961)
    at PaginationProxy.goToPage (ag-grid-community.js:20990)
    at PaginationProxy.goToNextPage (ag-grid-community.js:21060)
    at PaginationComp.onBtNext (ag-grid-community.js:40478)

**Comment by zhenyab** [comment]
> Which is the latest version that works?
> update: Works on v19, but there another bug with filtering state being reset. Every v21 is broken. Didn't check v20.

I do not know. I just started to use ag-grid.

**Comment by vicneanschi** [comment]
Same issue here with v21.2.1 Enterprise. Any updates?

**Comment by rdunce** [comment]
I submitted this a while ago with a plunkr(https://plnkr.co/edit/bRyIIHVltnSFBNRF3coS?p=preview) through their enterprise support, it's being tracked on ZenDesk AG-3322, it's related to issue AG-2659. Unfortunately it still seems to appear in the backlog. AgGrid 20.x.x works, none of the 21.x.x's do.

**Comment by DmitryEremenko** [comment]
My company considering to use enterprise version of this library but this bug not allowing to make decision. Is there any way to bypass this error?

**Comment by davis4acca** [comment]
Hi,

Thank you for your email. We are aware of this issue and have a feature request in our backlog:

You can follow the status of feature requests, bugs and releases using our pipeline:
https://www.ag-grid.com/ag-grid-pipeline/

If you are an ag-Grid Enterprise Customer, please raise this in our Zendesk support system - contact info@ag-grid.com for access. This is our primary channel for Support.



This is the reference and summary of the relevant issue:

AG-3322 | Regression 21.2.0- Server side row model + pagination + infinite scrolling does not work anymore for simple cases

**Comment by kiril-matev** [comment]
Hello everyone!

I'd like to update you on this issue. 

This issue was tracked by AG-3322, which  was resolved as part of ag-Grid v21.2.2 released on September 27th, 2019. If you're still seeing this issue please upgrade to v21.2.2 or later to resolve it.

If you're still seeing this issue with the latest version of ag-Grid, please enter the details as per the prescribed format so we can investigate.

Thank you for following this.
